src/cluster.c

```diff
@@ -1245,6 +1309,38 @@ void clusterRenameNode(clusterNode *node, char *newname) {
     clusterAddNode(node);
 }
 
+void clusterAddNodeToShard(const char *shard_id, clusterNode *node) {
+    sds s = sdsnewlen(shard_id, CLUSTER_NAMELEN);
+    dictEntry *de = dictFind(server.cluster->shards,s);
+    if (de == NULL) {
+        list *l = listCreate();
+        serverAssert(listAddNodeTail(l, node) == l);
+        serverAssert(dictAdd(server.cluster->shards, s, l) == DICT_OK);
+    } else {
+        list *l = dictGetVal(de);
+        if (listSearchKey(l, node) == NULL) {
+            serverAssert(listAddNodeTail(l, node) == l);
+        }
+        sdsfree(s);
+    }
+}
+
+void clusterRemoveNodeFromShard(const char *shard_id, clusterNode *node) {
+    sds s = sdsnewlen(shard_id, CLUSTER_NAMELEN);
+    dictEntry *de = dictFind(server.cluster->shards,s);
+    if (de != NULL) {
+        list *l = dictGetVal(de);
+        listNode *ln = listSearchKey(l, node);
+        if (ln != NULL) {
+            listDelNode(l, ln);
+        }
+        if (listLength(l) == 0) {
+            dictDelete(server.cluster->shards,s);
+        }
```

Comment:
I agree with you that there is value in allowing admin to control the shard id. The implementation should be straightforward. I am not sure though about the mechanism or interface that you have in mind. Are you suggesting a new cluster command like SETSHARDID or something else? FWIW, there is an odd way to reuse/customize the shard_id for a new node, which is through a manually crafted nodes.conf.

Reply:
How about a `JOINSHARD` primitive? I think it makes more sense to expose a high level construct around shards, as opposed to letting admins manipulate the shard-id directly. Specifically, I propose the below:

A node `N` can explicitly join a shard `S` as identified by its `<shard-id>` via `CLUSTER JOINSHARD <shard-id>` 
* If shard `S` is not empty, `N` will try locating its primary `P` and initiate replication, as long as it sees one live node in `S`
* otherwise, if all nodes in `S` are permanently down, `N` should attempt a manual failover
* otherwise, there is no node in `S` and this means that the cluster is not aware of this shard at all. I'd propose we fail the `JOINSHARD` request.

For completeness, the following two have been implemented in this PR:
1. A new node `N` always starts out in its own shard with an auto-generated shard-id
2. A node `N` implicitly joins an existing shard via `CLUSTER REPLICATE <primary-node-id>`

Reply:
Sounds good to me! It generally follows the paradigm of cluster v2 as well.

Reply:
@madolson what do you think about the logistics here? Do you prefer to having `JOINSHARD` implemented as part of this PR or you are open to a follow up PR?

Reply:
Followup is fine.

Reply:
created #10935 to track the work on `JOINSHARD`